pkg/apis/jenkins.io/v1/types_fact.go

```diff
@@ -0,0 +1,113 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:openapi-gen=true
+
+// Fact represents observed facts. Apps will generate Facts about the system.
+// A naming schema is required since each Fact has a name that's unique for the whole system.
+// Apps should prefix their generated Facts with the name of the App, like <app-name>-<fact>.
+// This makes that different Apps can't possibly have conflicting Fact names.
+//
+// For an app generating facts on a pipeline, which will be have several different executions, we recommend <app>-<fact>-<pipeline>.
+type Fact struct {
+	metav1.TypeMeta `json:",inline"`
+	// The Fact labels will be used to query the API for interesting Facts.
+	// The Apps responsible for creating Facts need to add the relevant labels.
+	// For example, creating Facts on a pipeline would create Facts with the following labels
+	// {
+	//   subjectkind: PipelineActivity
+	//   pipelineName: my-org-my-app-master-23
+	//   org: my-org
+	//   repo: my-app
+	//   branch: master
+	//   buildNumber: 23
+	// }
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	Spec   FactSpec   `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+	Status FactStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// FactSpec is the specification of a Fact
+type FactSpec struct {
+	Name             string        `json:"name" protobuf:"bytes,1,opt,name=name"`
```

Comment:
There should be a `ResourceReference` (see https://github.com/jenkins-x/jx/blob/master/pkg/apis/jenkins.io/v1/register.go#L76) that points to e.g. the `PipelineActivity` - this is the standard format used e.g. by Kubernetes ownership references. You should also consider adding an ownershop

Reply:
Would it work making `SubjectReference` field be a ResourceReference instead of string?

Reply:
yes that makes sense - what does `Subject` mean though?

Reply:
The thing that the Facts refer to. I can rename it to ResourceReference if you guys think that makes more sense.

Reply:
No it kinda makes sense to me :-)

Reply:
> We've decided to split this in two:
> 
> * First extract the CRD as it's
> * Improve the CRD applying our current knowledge
> 
> I believe the work for extraction is ready. Anything else that you folks think is required before merging?

Yes, the `ResourceReference` shouldn't be a string ;-)